src/adapter.rs

```diff
@@ -0,0 +1,461 @@
+#[cfg(all(target_os = "linux", feature = "bluetooth"))]
+use blurz::bluetooth_adapter::BluetoothAdapter as BluetoothAdapterBluez;
+#[cfg(all(target_os = "android", feature = "bluetooth"))]
+use blurdroid::bluetooth_adapter::Adapter as BluetoothAdapterAndroid;
+#[cfg(all(target_os = "macos", feature = "bluetooth"))]
+use blurmac::BluetoothAdapter as BluetoothAdapterMac;
+#[cfg(not(any(all(target_os = "linux", feature = "bluetooth"),
+              all(target_os = "android", feature = "bluetooth"),
+              all(target_os = "macos", feature = "bluetooth"))))]
+use empty::BluetoothAdapter as BluetoothAdapterEmpty;
+#[cfg(feature = "bluetooth-test")]
+use blurmock::fake_adapter::FakeBluetoothAdapter;
+
+#[cfg(feature = "bluetooth-test")]
+const NOT_SUPPORTED_ON_REAL_ERROR: &'static str = "Error! Test functions are not supported on real devices!";
+#[cfg(feature = "bluetooth-test")]
+const NOT_SUPPORTED_ON_MOCK_ERROR: &'static str = "Error! The first parameter must be a mock structure!";
+
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::error::Error;
+use bluetooth::BluetoothDiscoverySession;
+use bluetooth::BluetoothDevice;
+
+pub trait BluetoothAdapter {
+    fn new()-> Result<Box<BluetoothAdapter>, Box<Error>>;
```

Comment:
This `new` method cannot be part of the trait, since it causes a compile error. Instead use the solution I described in a previous email.

Reply:
From my understanding of the email, I changed the init() functions to new() as your solution had the following,
>impl BluetoothAdapter {
>#[cfg(all(target_os = "macos", feature = "bluetooth"))]
>  pub fn new() -> Result<Box<BluetoothAdapter>, Box<Error>> {
>    // ...
> }
>This will allow callers to use `BluetoothAdapter::new()` to create a new 
 >instance.

Reply:
The code snippet I wrote in the email is different than what you have here. The snippet implements a bunch of per-platform static methods named `new` on all `BluetoothAdapter` trait objects. The code in this PR adds a static `new` method to the `BluetoothAdapter` trait, which then must be part of each concrete implementation of the trait, and the compiler does not like that.

Reply:
Hi Josh,
We modified the code as per your suggestion and we're still facing some issues.
One of the errors that we are getting is, 'no associated item named `Empty` found for type `dyn adapter::BluetoothAdapter` in the current scope'.  (Compiling on Windows).

impl BluetoothAdapter {
#[cfg(not(any(all(target_os = "linux", feature = "bluetooth"),
                  all(target_os = "android", feature = "bluetooth"),
                  all(target_os = "macos", feature = "bluetooth"))))]
    pub fn new() -> Result<Box<BluetoothAdapter>, Box<Error>> {
        let adapter = try!(BluetoothAdapterEmpty::init());
        Ok(BluetoothAdapter::Empty(Arc::new(adapter)))
    }
}
Could you give suggestions to fix this issue?

Reply:
You need to use `Ok(Box::new(adapter))` to return the type that is required.